### PR TITLE
[WIP] Windows build

### PR DESCRIPTION
**Pull request by bilderbuchi**
Checklist
* [x] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [x] Bumped the build number (if the version is unchanged)
* [x] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [ ] Ensured the license file is being packaged.
* [ ] Try to use `- {{ compiler('m2w64_cxx') }} # [win]` again and re-render the recipe afterwards!

Let's see how far we get in adding a Windows build of gdb.
See #6.

### CODE REVIEW

**Comment by bilderbuchi** [comment]
@conda-forge-admin , please rerender

**Comment by bilderbuchi** [comment]
um, I'm getting up to the `make` stage now, but I get "configure: error: C compiler cannot create executables" ([link](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=337460&view=logs&j=5be07ae1-d8ba-5406-47b6-8e3a3a12f825&t=0bf03e01-0bec-5b85-5316-b1633322e895&l=625)) 
No idea what to do about that...

**Comment by bilderbuchi** [comment]
@conda-forge-admin, please rerender

**Comment by bilderbuchi** [comment]
Hm, this [fails](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=345556&view=logs&j=171a126d-c574-5c8c-1269-ff3b989e923d&t=1183ba29-a0b5-5324-8463-2a49ace9e213) with 
```
conda.exceptions.ResolvePackageNotFound: 
  - ncurses
```
although I [specifically replaced ncurses by m2-ncurses](https://github.com/conda-forge/gdb-feedstock/pull/42/files#diff-f3725a55bf339595bf865fec73bda8ac99f283b0810c205442021f29c06eea9aR34-R35).

I can't recognize where/why it needs ncurses. The [log output before](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=345556&view=logs&j=171a126d-c574-5c8c-1269-ff3b989e923d&t=1183ba29-a0b5-5324-8463-2a49ace9e213&l=55) even seems to suggest that conda's finished, but then *another* solve fails?!

Curiously enough, this did not occur in the previous incarnation of this PR. Maybe a change in the recipe rerender template?

**Comment by bilderbuchi** [comment]
@klauer: interestingly, the re-render above (181521f) apparently broke (only) the [macos py39 build](https://travis-ci.com/github/conda-forge/gdb-feedstock/jobs/521003267), although no mac files seems to have been touched?

**Comment by klauer** [comment]
Strange - diffing the working build output to the non-working one, I don't see many differences:
This PR's build:
* `conda-forge-ci-setup-3.9.4`
* Test phase fails with the gdb output `During startup program terminated with signal ?, Unknown signal.` which happens before even a thread gets created (*)
* Doesn't touch any of the travis config as you noted

<details>
<summary>(*) Test diff</summary>

Working:
```
Developer mode is already enabled.
YES (0)
[New Thread 0x1503 of process 40618]
[New Thread 0x2803 of process 40618]
warning: `/BuildRoot/Library/Caches/com.apple.xbs/Binaries/Libc_darwin/install/TempContent/Objects/Libc.build/libsystem_darwin.dylib.build/Objects-normal/x86_64/darwin_vers.o': can't open to read symbols: No such file or directory.
warning: `/BuildRoot/Library/Caches/com.apple.xbs/Binaries/Libc_darwin/install/TempContent/Objects/Libc.build/libsystem_darwin.dylib.build/Objects-normal/x86_64/dirstat.o': can't open to read symbols: No such file or directory.
warning: `/BuildRoot/Library/Caches/com.apple.xbs/Binaries/Libc_darwin/install/TempContent/Objects/Libc.build/libsystem_darwin.dylib.build/Objects-normal/x86_64/dirstat_collection.o': can't open to read symbols: No such file or directory.
warning: `/BuildRoot/Library/Caches/com.apple.xbs/Binaries/Libc_darwin/install/TempContent/Objects/Libc.build/libsystem_darwin.dylib.build/Objects-normal/x86_64/init.o': can't open to read symbols: No such file or directory.
warning: `/BuildRoot/Library/Caches/com.apple.xbs/Binaries/Libc_darwin/install/TempContent/Objects/Libc.build/libsystem_darwin.dylib.build/Objects-normal/x86_64/variant.o': can't open to read symbols: No such file or directory.
Hello, World!
[Inferior 1 (process 40618) exited normally]
```

Not working:
```
Developer mode is already enabled.
YES (0)
During startup program terminated with signal ?, Unknown signal.
[New Thread 0x1503 of process 40602]
Tests failed for gdb-10.2-py39h593ab31_0.tar.bz2 - moving package to /Users/travis/miniforge3/conda-bld/broken
WARNING:conda_build.build:Tests failed for gdb-10.2-py39h593ab31_0.tar.bz2 - moving package to /Users/travis/miniforge3/conda-bld/broken
WARNING conda_build.build:tests_failed(2955): Tests failed for gdb-10.2-py39h593ab31_0.tar.bz2 - moving package to /Users/travis/miniforge3/conda-bld/broken
```
</details>

Working build
* `conda-forge-ci-setup-3.9.5` (no significant differences as far as I can tell, [recipe here](https://github.com/conda-forge/conda-forge-ci-setup-feedstock/commits/master/recipe/meta.yaml))
* Everything else looks the same

I'd imagine the tests aren't so flaky as the 10.2 PR merged and built without [issue](https://travis-ci.com/github/conda-forge/gdb-feedstock/jobs/520973280)...

**Comment by bilderbuchi** [comment]
Curiously, on the next run, both macos-py36 and macos-py39 [failed](https://github.com/conda-forge/gdb-feedstock/runs/2966213305), although again no macos config was touched. Maybe something is flaky after all. 

Btw, easter-egg time, gdb still contains python 2 code, although that [does not fail the build](https://travis-ci.com/github/conda-forge/gdb-feedstock/jobs/521007149#L6082) 😁

**Comment by bilderbuchi** [comment]
Meanwhile, there's conda-forge-ci-setup 3.9.*6* available, so
@conda-forge-admin, please rerender

**Comment by phil-blain** [comment]
I'll try to help if I can for the Windows build, but I have close to zero experience on Windows. 

Regarding the macOS Travis run that seems flaky, upon re-reading the GDB bugzilla entry linked in [recipe/README](../blob/master/recipe/README.md), I think that I would have to update the patch that we use for macOS. I'll try to do that some time. In the meantime I think we can ignore macOS failures in this here PR,  at least the `During startup program terminated with signal ?, Unknown signal.`  error.

Regarding the latest macOS py36 failure, I see this at the end:
~~~
Solving environment: ...working... failed
failed to get install actions, retrying.  exception was: Unsatisfiable dependencies for platform osx-64: {'gdb==10.2=py36h7a1bca3_1'}
WARNING:conda_build.build:failed to get install actions, retrying.  exception was: Unsatisfiable dependencies for platform osx-64: {'gdb==10.2=py36h7a1bca3_1'}
WARNING conda_build.build:test(2859): failed to get install actions, retrying.  exception was: Unsatisfiable dependencies for platform osx-64: {'gdb==10.2=py36h7a1bca3_1'}
~~~

I'm not sure about this one. I don't see anything wrong before that. There's a lot of "is not an ELF" lines but that might just be conda-build acting out, I'm not sure. It's the same thing for the macOS py39 failure in 181521fa05021dd4e5b4dcebaeb08f4c0e42ba85...

**Comment by bilderbuchi** [comment]
At this point, during running the make target "intl-configure", I get
```
checking for x86_64-pc-mingw64-gcc... cl.exe
checking whether the C compiler works... no
configure: error: in `/d/bld/gdb_1625476257175/work/build/intl':
configure: error: C compiler cannot create executables
```
I suspect the problem is that the MSVC compiler (cl.exe) gets picked up before/instead of gcc, and gdb just can't use that one. Maybe we can't use the automated `{{ compiler('c') }}` requirement, and need to specifically prescribe gcc (which package? `m2-*`? `m2w64-*`?)

**Comment by bilderbuchi** [comment]
I don't know why the azure jobs were canceled. Can somebody with permissions re-trigger them, maybe?

**Comment by bilderbuchi** [comment]
OK, I give up. I'm out of things to try. Obviously I'm not knowledgeable enough to pull this off. :-(

**Comment by bilderbuchi** [comment]
@conda-forge-admin restart ci

**Comment by bilderbuchi** [comment]
I'll try one more thing - a probably more correct compiler name on Windows. 
I found another poor soul (@WardBrian) wading through the same seas - thanks for the inspiration! One difference - I'm trying to run the `build.sh` on Windows, too, and `bld.bat` is just a smallish shim, maybe that could be useful for you?

**Comment by phil-blain** [comment]
Hi @bilderbuchi ! I think that `m2w64_cxx` is the right direction. But the fact is that the m2w64 ecosystem in conda-forge is not very developed. Most of the package are quite old. I'm not even even sure if the packages that are available were built in the conda-forge infrastructure, or were just repackaged from the defaults (i.e. Anaconda) "msys" channel. The conda-forge org on GitHub has just 4 such packages: https://github.com/conda-forge?q=m2w64&type=&language=&sort= but more are available with `conda search`, so I'm not sure.

 I think that to get more help with a Windows build of GDB, or know more about the current state of the m2w64 conda-forge ecosystem, it would help to get help from members of conda-forge core, either by pinging them here (`@conda-forge/core`) or reaching out on Gitter. There is some discussion in this issue: https://github.com/conda-forge/conda-forge.github.io/issues/1044.

**Comment by bilderbuchi** [comment]
Thanks for the pointers! Turns out I'm already subscribed to that issue, I'd also welcome a more modern toolchain being available! :-)

In the meantime, hard-coding the compiler name instead of letting conda build figure it out brought progress!
I'm seeing two issues:
`cmp: command not found` (I'll have to find out which package provides that)

This snippet suggests to me that a missing `makeinfo` tool fails a target `bfd.info`, which subsequently fails the `all` target and thus, the build.
```
2021-07-22T07:26:56.1197916Z rm -rf $backupdir; exit $rc
2021-07-22T07:26:56.3147928Z /d/bld/gdb_1626937951290/work/missing: line 81: makeinfo: command not found
2021-07-22T07:26:56.3714549Z WARNING: 'makeinfo' is missing on your system.
2021-07-22T07:26:56.3736914Z          You should only need it if you modified a '.texi' file, or
2021-07-22T07:26:56.3737925Z          any other file indirectly affecting the aspect of the manual.
2021-07-22T07:26:56.3738946Z          You might want to install the Texinfo package:
2021-07-22T07:26:56.3739633Z          <http://www.gnu.org/software/texinfo/>
2021-07-22T07:26:56.3740364Z          The spurious makeinfo call might also be the consequence of
2021-07-22T07:26:56.3741248Z          using a buggy 'make' (AIX, DU, IRIX), in which case you might
2021-07-22T07:26:56.3741882Z          want to install GNU make:
2021-07-22T07:26:56.3742641Z          <http://www.gnu.org/software/make/>
2021-07-22T07:26:56.4343316Z make[3]: *** [bfd.info] Error 127
2021-07-22T07:26:56.4344093Z Makefile:542: recipe for target 'bfd.info' failed
2021-07-22T07:26:56.4344621Z make[3]: Leaving directory '/d/bld/gdb_1626937951290/work/build/bfd/doc'
2021-07-22T07:26:56.4378507Z make[2]: *** [info-recursive] Error 1
2021-07-22T07:26:56.4379092Z Makefile:1643: recipe for target 'info-recursive' failed
2021-07-22T07:26:56.4379764Z make[2]: Leaving directory '/d/bld/gdb_1626937951290/work/build/bfd'
2021-07-22T07:26:56.4450178Z make[1]: *** [all-bfd] Error 2
2021-07-22T07:26:56.4450909Z Makefile:2777: recipe for target 'all-bfd' failed
2021-07-22T07:26:56.4454528Z make[1]: *** Waiting for unfinished jobs....
2021-07-22T07:26:56.7767370Z no
```

**Comment by bilderbuchi** [comment]
Ah that's great. :sarcasm:
So, gdb fails the make stage if `makeinfo` is not installed. Reported regularly:  
https://sourceware.org/bugzilla/show_bug.cgi?id=14678
https://sourceware.org/bugzilla/show_bug.cgi?id=18113
https://sourceware.org/bugzilla/show_bug.cgi?id=20417

Apparently there is no switch to convince gdb to correctly drop building the documentation

`Makeinfo` seems to be part of `texinfo`, which even has a [c-f package](https://github.com/conda-forge/texinfo-feedstock). Unfortunately, no Windows builds! 😭 

apparently, the options are
* somehow manually insert a `GDBvn.texi` so that make does not try to use makeinfo to rebuild `gdb.info`.
* create a faked makeinfo tool in the right place that just emits a version string, so that the configure step runs properly(?), see https://sourceware.org/bugzilla/show_bug.cgi?id=18113#c1.

**Comment by phil-blain** [comment]
There is a m2-texinfo package on the msys2 channel (which I think is copied in conda-forge) https://anaconda.org/msys2/m2-texinfo. 

P.s. the official GDB patchwork instance is at https://patchwork.sourceware.org/project/gdb/list/ ;)

**Comment by bilderbuchi** [comment]
Ah, good to know -- I did not realise because I used my mambaforge to search, which has conda-forge only. 

Thanks for the official link. While I did not find the second patch there, it seems to have [landed in master](https://sourceware.org/git/?p=binutils-gdb.git;a=commit;h=ab954e4a53cab8c33728e6708695abc1616c90b1) back in December because I see the [changelog entry in master](https://sourceware.org/git/?p=binutils-gdb.git;a=blob;f=gdb/doc/ChangeLog;h=0df8399caa30207ff13246f322b0818ef29059df;hb=refs/heads/master)

**Comment by bilderbuchi** [comment]
Now it fails in something called "dwarf2" with `bfd/dwarf2.c:808:71: error: incomplete universal character name \u` 
No idea about this at this point...

Every day a little farther, this starts to feel like thesis work :-P

**Comment by phil-blain** [comment]
I checked the MSYS package and it actually carries some patches: https://github.com/msys2/MINGW-packages/tree/master/mingw-w64-gdb. Maybe checking in some of these might help... 

Cygwin also carries a few patches: https://cygwin.com/git-cygwin-packages/?p=git/cygwin-packages/gdb.git;a=tree

The MinGW-w64 builds [here](https://sourceforge.net/projects/mingw-w64/files/Toolchains%20targetting%20Win64/Personal%20Builds/mingw-builds/) also provide GDB, but I'm not sure whre is the source repo for that (Sourceforge UI is so messy :o)

**Comment by bilderbuchi** [comment]
From the other warnings surrouding this, I suspect that this could be some remaining Windows-style paths that trip up this function.

**Comment by maresb** [comment]
@bilderbuchi, I noticed your mention of https://github.com/conda-forge/staged-recipes/pull/15203. I have no idea if it's helpful for you, but @WardBrian got things working a few days ago.

**Comment by bilderbuchi** [comment]
yeah, I think we both hit a temporary CI breakdown, which has since resolved. I managed to get over some more roadblocks, and checked out the changes of that PR, but at this point I think my problems have a different source (`\` in windows paths tripping up this dwarf function), but it's hard to troubleshoot.

**Comment by bilderbuchi** [comment]
Specifically, I suspect I need to add one more env vars to the cygpath conversion done [here](https://github.com/conda-forge/gdb-feedstock/pull/42/files#diff-54cf74e113dd3f6d11e092fdb1d888ec82c69bdafbb15cfb6570c83ecad28f33R2-R9), but I currently don't know how to find out *which* path is the culprit. I think the warnings about escapes surrounding the [error linked above](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=355207&view=logs&j=e5cdccbf-4751-5a24-7406-185c9d30d021&t=66e4a587-a723-53e9-873c-6c9a2c5971ea&l=3561) could each stem from the initial letters of directories in the affected path, but I have not yet tried to decipher that.

**Comment by bilderbuchi** [comment]
The `DEBUGDIR` path in the lines [1](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=355207&view=logs&j=e5cdccbf-4751-5a24-7406-185c9d30d021&t=66e4a587-a723-53e9-873c-6c9a2c5971ea&l=3546) and [2](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=355207&view=logs&j=e5cdccbf-4751-5a24-7406-185c9d30d021&t=66e4a587-a723-53e9-873c-6c9a2c5971ea&l=3548) (and what follows in the same line is suspiciously similar to the pattern of invalid escape sequence warnings. 
I suspect that this somehow (how?!) mangles forward-slash delimited paths as in the command line invocation into backslash delimited paths (with some expansion as a capital `\L` is also in there?), which then trigger all these invalid escape sequence warnings?

**Comment by bilderbuchi** [comment]
OK, now it [fails](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=366228&view=logs&j=e5cdccbf-4751-5a24-7406-185c9d30d021&t=66e4a587-a723-53e9-873c-6c9a2c5971ea&l=3887) because it does not find ` netdb.h`, which is apparently not included in mingw-64:
> D:/bld/gdb_1629543547423/work/gdbsupport/netstuff.cc:28:19: fatal error: netdb.h: No such file or directory

Curiously enough, [above that point](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=366228&view=logs&j=e5cdccbf-4751-5a24-7406-185c9d30d021&t=66e4a587-a723-53e9-873c-6c9a2c5971ea&l=3646) it checks for the presence of netdb.h and correctly recognizes that it's not available, so I don't know why it stumbles over that later.

Note that the tests are *not successul*, but Azure apparently currently does not report back results.

**Comment by bilderbuchi** [comment]
Since gdb is not available from conda-forge for Windows, and it seems to me it should be, I have been trying to get that to build.
Apparently, I don't know what I'm doing, because I've been banging my head against this over the summer. At this point I'm (again) stuck.

Maybe some Windows/m2w64/mingw/gcc experts can take a look and help out getting gdb built? @conda-forge/help-c-cpp @conda-forge/core

**Comment by bilderbuchi** [comment]
My reasoning for going the route of a slim bld.bat shim that calls build.sh was to minimize the additional maintenance effort for the Windows build by not introducing _another_ build script. I saw that approach in another package, so it seemed reasonable.

**Comment by jaimergp** [comment]
Looks like a missing header: [`netdb.h`](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=366228&view=logs&j=171a126d-c574-5c8c-1269-ff3b989e923d&t=1183ba29-a0b5-5324-8463-2a49ace9e213&l=3884). This seems to be _not_ part of MingW, so you'll need to patch it to replace it by a Windows equivalent. [Source 1](https://github.com/msys2/MINGW-packages/issues/3586), [source 2](https://github.com/idris-lang/Idris-dev/issues/3732#issuecomment-291248450), [source 3](https://stackoverflow.com/questions/4432989/c-header-files-for-udp-in-windows/4433031).

File: recipe/meta.yaml [hunk 1]
```diff
@@ -12,35 +12,44 @@ source:
...
 
 requirements:
   build:
     - {{ compiler('fortran') }}
-    - {{ compiler('c') }}
-    - {{ compiler('cxx') }}
-    - make >=3.82
+    - {{ compiler('c') }}          # [not win]
+    - {{ compiler('cxx') }}        # [not win]
+    - m2w64-gcc                    # [win]
```
Comment (jaimergp) [inline recipe/meta.yaml:25]: I think there's a `compiler` function for m2w64 too. See [here](https://github.com/conda-forge/ambertools-feedstock/pull/34/files#diff-f3725a55bf339595bf865fec73bda8ac99f283b0810c205442021f29c06eea9aR34-R36).

Comment (jaimergp) [reply recipe/meta.yaml:25]: You need to rerender the recipe for this to have an effect.

**Comment by bilderbuchi** [comment]
> It looks like configure should handle this but it does not. 

Yeah, right? That's what I thought, too. That `USE_WIN32API` thing looks promising!

It's funny, mingw-64 actually *has* gdb available (in fact, that's what I'm using in the meantime, but not installed via the conda ecosystem), I just can't find out how the manage to build it... :-/

**Comment by bilderbuchi** [comment]
Thanks for those tips! This is the relevant log section
```
2021-08-29T20:44:46.4356997Z g++  -std=gnu++17    -I. -I/d/bld/gdb_1630268986132/work/gdbserver -I/d/bld/gdb_1630268986132/work/gdbserver/../gdb/regformats -I/d/bld/gdb_1630268986132/work/gdbserver/.. -I/d/bld/gdb_1630268986132/work/gdbserver/../include -I/d/bld/gdb_1630268986132/work/gdbserver/../gdb -I/d/bld/gdb_1630268986132/work/gdbserver/../gnulib/import -I../gnulib/import -I/d/bld/gdb_1630268986132/work/gdbserver/.. -I.. -I./../intl -I/d/bld/gdb_1630268986132/_h_env/include -DUSE_WIN32API -pthread -Wall -Wpointer-arith -Wno-unused -Wunused-value -Wunused-variable -Wunused-function -Wno-switch -Wno-char-subscripts -Wempty-body -Wunused-but-set-parameter -Wunused-but-set-variable -Wno-sign-compare -Wno-error=maybe-uninitialized -Wsuggest-override -Wmissing-declarations -Wstrict-null-sentinel -Wformat -Wformat-nonliteral  -DGDBSERVER -lwsock32 -lws2_32 -Wl,--stack,12582912  \
2021-08-29T20:44:46.4358761Z 	-o gdbreplay.exe gdbreplay.o utils.o version.o  \
2021-08-29T20:44:46.4359436Z 	../gdbsupport/libgdbsupport.a ../gnulib/import/libgnu.a ../libiberty/libiberty.a ./../intl/libintl.a -liconv \
2021-08-29T20:44:46.4360053Z 	
2021-08-29T20:44:46.4360641Z gdbreplay.o:gdbreplay.cc:(.text+0xf7): undefined reference to `__imp_closesocket'
2021-08-29T20:44:46.4361495Z gdbreplay.o:gdbreplay.cc:(.text+0x224): undefined reference to `__imp_WSAStartup'
2021-08-29T20:44:46.4362202Z gdbreplay.o:gdbreplay.cc:(.text+0x276): undefined reference to `__imp_getaddrinfo'
2021-08-29T20:44:46.4363098Z gdbreplay.o:gdbreplay.cc:(.text+0x294): undefined reference to `gai_strerrorA'
2021-08-29T20:44:46.4363768Z gdbreplay.o:gdbreplay.cc:(.text+0x35d): undefined reference to `__imp_socket'
2021-08-29T20:44:46.4364492Z gdbreplay.o:gdbreplay.cc:(.text+0x3d3): undefined reference to `__imp_setsockopt'
2021-08-29T20:44:46.4365190Z gdbreplay.o:gdbreplay.cc:(.text+0x482): undefined reference to `__imp_bind'
2021-08-29T20:44:46.4365855Z gdbreplay.o:gdbreplay.cc:(.text+0x4db): undefined reference to `__imp_listen'
2021-08-29T20:44:46.4366680Z gdbreplay.o:gdbreplay.cc:(.text+0x511): undefined reference to `__imp_accept'
2021-08-29T20:44:46.4367391Z gdbreplay.o:gdbreplay.cc:(.text+0x56a): undefined reference to `__imp_setsockopt'
2021-08-29T20:44:46.4368068Z gdbreplay.o:gdbreplay.cc:(.text+0x5a6): undefined reference to `__imp_setsockopt'
2021-08-29T20:44:46.4368789Z gdbreplay.o:gdbreplay.cc:(.text+0x5e2): undefined reference to `__imp_getnameinfo'
2021-08-29T20:44:46.4369609Z gdbreplay.o:gdbreplay.cc:(.text+0x641): undefined reference to `__imp_closesocket'
2021-08-29T20:44:46.4370351Z gdbreplay.o:gdbreplay.cc:(.rdata$.refptr.in6addr_any[.refptr.in6addr_any]+0x0): undefined reference to `in6addr_any'
2021-08-29T20:44:46.4371166Z ../gdbsupport/libgdbsupport.a(netstuff.o):netstuff.cc:(.text+0x57): undefined reference to `__imp_freeaddrinfo'
2021-08-29T20:44:46.4371878Z collect2.exe: error: ld returned 1 exit status
2021-08-29T20:44:46.4372503Z Makefile:362: recipe for target 'gdbreplay.exe' failed
2021-08-29T20:44:46.4373101Z make[2]: *** [gdbreplay.exe] Error 1
```
gdbreplay.o gets compiled here (apparently without mishap)
```
2021-08-29T20:44:28.5697995Z g++  -std=gnu++17    -I. -I/d/bld/gdb_1630268986132/work/gdbserver -I/d/bld/gdb_1630268986132/work/gdbserver/../gdb/regformats -I/d/bld/gdb_1630268986132/work/gdbserver/.. -I/d/bld/gdb_1630268986132/work/gdbserver/../include -I/d/bld/gdb_1630268986132/work/gdbserver/../gdb -I/d/bld/gdb_1630268986132/work/gdbserver/../gnulib/import -I../gnulib/import -I/d/bld/gdb_1630268986132/work/gdbserver/.. -I.. -I./../intl -I/d/bld/gdb_1630268986132/_h_env/include -DUSE_WIN32API -pthread -Wall -Wpointer-arith -Wno-unused -Wunused-value -Wunused-variable -Wunused-function -Wno-switch -Wno-char-subscripts -Wempty-body -Wunused-but-set-parameter -Wunused-but-set-variable -Wno-sign-compare -Wno-error=maybe-uninitialized -Wsuggest-override -Wmissing-declarations -Wstrict-null-sentinel -Wformat -Wformat-nonliteral  -DGDBSERVER -c -o gdbreplay.o -MT gdbreplay.o -MMD -MP -MF ./.deps/gdbreplay.Tpo /d/bld/gdb_1630268986132/work/gdbserver/gdbreplay.cc
```
The missing includes (WSAStartup etc) referenced above are behind `#ifdef USE_WIN32API` guards in ` gdbreplay.c`, so the `-DUSE_WIN32API` seems to work in order. It just cannot seem to find the library linked with ` -lwsock32 -lws2_32`, I don't know why.

**Comment by phil-blain** [comment]
Copying the linking command with some newlines and commands added:
~~~shell
g++  -std=gnu++17    \
# Includes
-I. -I/d/bld/gdb_1630268986132/work/gdbserver \
-I/d/bld/gdb_1630268986132/work/gdbserver/../gdb/regformats \
-I/d/bld/gdb_1630268986132/work/gdbserver/.. \
-I/d/bld/gdb_1630268986132/work/gdbserver/../include \
-I/d/bld/gdb_1630268986132/work/gdbserver/../gdb \
-I/d/bld/gdb_1630268986132/work/gdbserver/../gnulib/import \
-I../gnulib/import -I/d/bld/gdb_1630268986132/work/gdbserver/.. \
-I.. -I./../intl -I/d/bld/gdb_1630268986132/_h_env/include \
# Preprocessor
-DUSE_WIN32API -pthread \
# Warnings
-Wall -Wpointer-arith -Wno-unused -Wunused-value -Wunused-variable -Wunused-function -Wno-switch -Wno-char-subscripts -Wempty-body -Wunused-but-set-parameter -Wunused-but-set-variable -Wno-sign-compare -Wno-error=maybe-uninitialized -Wsuggest-override -Wmissing-declarations -Wstrict-null-sentinel -Wformat -Wformat-nonliteral  \
# More preprocessor, your added LDFLAGS
-DGDBSERVER -lwsock32 -lws2_32 -Wl,--stack,12582912  \
# Object files
-o gdbreplay.exe gdbreplay.o utils.o version.o  \
# Static libs
../gdbsupport/libgdbsupport.a ../gnulib/import/libgnu.a ../libiberty/libiberty.a ./../intl/libintl.a \
# Shared libs
-liconv 	
~~~

According to [this SO answer](https://stackoverflow.com/a/2033632), the `-lwsock32 -lws2_32` have to come *after* the object files. I think that's what you'd want to achieve, somehow. 

Looking briefly at the code, I think this *should* all work correctly, without having to set `USE_WIN32API` manually.
Here is an excerpt from `gdb/configure`:
~~~shell
  WIN32APILIBS=
  case ${host} in
    *mingw32*)

$as_echo "#define USE_WIN32API 1" >>confdefs.h

      WIN32APILIBS="-lws2_32"
      ;;
  esac
~~~

So I think the problem is that `host` is not what this configure bit expects, in fact configure says it's: [`x86_64-pc-mingw64`](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=369351&view=logs&j=171a126d-c574-5c8c-1269-ff3b989e923d&t=1183ba29-a0b5-5324-8463-2a49ace9e213&l=513) (which I think is correct). So maybe a simple fix would be to patch this configure (and maybe others) to use `*mingw*` instead of `*mingw32*` ?

**Comment by bilderbuchi** [comment]
I'm closing this, this is hopelessly stale by now. Someone else, feel free to pick up where I left.

**Pull request closed without merging**